### PR TITLE
Cleanup custom data storage

### PR DESCRIPTION
**Pull request by jonnybbb**
After switching to a new storage system for build scan data in [version 0.12]( https://github.com/etiennestuder/teamcity-build-scan-plugin/tree/v0.12) m the previously used custom data storage can be cleaned up to free up space.

### CODE REVIEW

File: src/main/resources/buildServerResources/buildScanAdminPage.jsp [hunk 1]
```diff
@@ -0,0 +1,18 @@
+<%@taglib prefix="c" uri="http://java.sun.com/jsp/jstl/core" %>
+    <%@include file="/include-internal.jsp" %>
+    <jsp:useBean id="count" type="java.lang.Integer" scope="request"/>
+    <h3>Cleanup storage of previous plugin versions</h3>
+    <c:choose>
+        <c:when test = "${count > 0}">
+            <h4>Warning: This will remove all build scans from previous versions of this plugin</h4>
+            <authz:authorize allPermissions="CONFIGURE_SERVER_DATA_CLEANUP">
+                <div id="cleanupResult"></div><div id="showCleanupProcessRunning" style="display:none;"><forms:progressRing className="progressRingInline"/></div>
```
Comment (jonnybbb) [inline src/main/resources/buildServerResources/buildScanAdminPage.jsp:9]: This UI experience can be improved if it should stay.

File: src/main/java/nu/studer/teamcity/buildscan/internal/cleanup/CustomDataStorageBuildScanCleaner.java [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+import java.sql.*;
+import java.util.ArrayList;
+import java.util.List;
+
+import static nu.studer.teamcity.buildscan.internal.CustomDataStorageBuildScanDataStore.BUILD_SCAN_STORAGE_ID;
+
+@SuppressWarnings("deprecation")
+public final class CustomDataStorageBuildScanCleaner {
+
+    private static final int BATCH_SIZE = 10;
```
Comment (etiennestuder) [inline src/main/java/nu/studer/teamcity/buildscan/internal/cleanup/CustomDataStorageBuildScanCleaner.java:15]: This seems a very low number, though might very well be justified. Can you please elaborate?

File: src/main/java/nu/studer/teamcity/buildscan/internal/cleanup/BuildScanAdminPage.java [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+public class BuildScanAdminPage extends AdminPage {
+
+    private final CustomDataStorageBuildScanCleaner customDataStorageBuildScanCleaner;
+
+    public BuildScanAdminPage(@NotNull PagePlaces pagePlaces, @NotNull PluginDescriptor descriptor, CustomDataStorageBuildScanCleaner customDataStorageBuildScanCleaner) {
+        super(pagePlaces);
+        this.customDataStorageBuildScanCleaner = customDataStorageBuildScanCleaner;
+        setPluginName(descriptor.getPluginName());
+        setIncludeUrl(descriptor.getPluginResourcesPath("/admin/buildScanAdminPage.jsp"));
+        addJsFile(descriptor.getPluginResourcesPath("js/buildScan.js"));
```
Comment (etiennestuder) [inline src/main/java/nu/studer/teamcity/buildscan/internal/cleanup/BuildScanAdminPage.java:22]: Would `/js/buildScan.js` work, too?

File: src/main/java/nu/studer/teamcity/buildscan/internal/cleanup/BuildScanAdminPage.java [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+        setTabTitle("Build Scan");
+        setPosition(PositionConstraint.last());
+        register();
```
Comment (etiennestuder) [inline src/main/java/nu/studer/teamcity/buildscan/internal/cleanup/BuildScanAdminPage.java:25]: I suggest to replace this with the `init` configuration on the bean in the Spring XML file.

Comment (etiennestuder) [reply src/main/java/nu/studer/teamcity/buildscan/internal/cleanup/BuildScanAdminPage.java:25]: Done

File: src/main/resources/buildServerResources/admin/buildScanAdminPage.jsp [hunk 5]
```diff
@@ -0,0 +1,23 @@
+<%@taglib prefix="c" uri="http://java.sun.com/jsp/jstl/core" %>
+    <%@include file="/include-internal.jsp" %>
+    <jsp:useBean id="count" type="java.lang.Integer" scope="request"/>
+    <h3>Cleanup storage of previous plugin versions</h3>
```
Comment (etiennestuder) [inline src/main/resources/buildServerResources/admin/buildScanAdminPage.jsp:4]: Should be same style/size as on the plugins list page

File: src/main/java/nu/studer/teamcity/buildscan/internal/cleanup/BuildScanAdminPage.java (hunk 3 above)
Comment (etiennestuder) [inline src/main/java/nu/studer/teamcity/buildscan/internal/cleanup/BuildScanAdminPage.java:22]: Can you please move this under /admin as well? (same folder as buildScanAdminPage.jsp)

File: src/main/java/nu/studer/teamcity/buildscan/internal/cleanup/BuildScanCleanupController.java [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+    private final SecurityContext securityContext;
+
+    public BuildScanCleanupController(@NotNull WebControllerManager controllerManager,
+                                      @NotNull CustomDataStorageBuildScanCleaner customDataStorageCleaner,
+                                      @NotNull SecurityContext securityContext) {
+        super(controllerManager);
+        this.customDataStorageCleaner = customDataStorageCleaner;
+        this.securityContext = securityContext;
+        controllerManager.registerController("/admin/buildScanCleanup.html", this);
+        this.registerAction(this);
```
Comment (etiennestuder) [inline src/main/java/nu/studer/teamcity/buildscan/internal/cleanup/BuildScanCleanupController.java:28]: Just for my understanding: why is registering the controller not enough, i.e. why is registerAction also needed?

Comment (etiennestuder) [reply src/main/java/nu/studer/teamcity/buildscan/internal/cleanup/BuildScanCleanupController.java:28]: Looking at the Spring code, I now understand.

File: src/main/resources/buildServerResources/js/buildScan.js [hunk 7]
```diff
@@ -0,0 +1,23 @@
+BS.BuildScan = {
+    startCleanup: function () {
+        if (!confirm("This operation may require significant time, and the server will be less performant during the clean-up.\n" +
+            "Are you sure you want to start clean-up process now?")) return false;
+        $('cleanupResult').style.visibility = 'visible'
```
Comment (etiennestuder) [inline src/main/resources/buildServerResources/js/buildScan.js:5]: Let's remove the confirmation dialog

File: src/main/resources/buildServerResources/admin/buildScanAdminPage.jsp [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+    <c:choose>
+        <c:when test = "${count > 0}">
+            <h4>Warning: This will remove all build scans from previous versions of this plugin</h4>
+            <authz:authorize allPermissions="CONFIGURE_SERVER_DATA_CLEANUP">
+
+                <table>
+                  <tr><td><div id="cleanupResult" style="visibility:hidden;">Not running</div></td><td></td></tr>
+                  <tr><td><form id="cleanupCustomDataStorage" method="POST" action="<c:url value='/admin/buildScanCleanup.html'/>" onsubmit="">
```
Comment (etiennestuder) [inline src/main/resources/buildServerResources/admin/buildScanAdminPage.jsp:12]: Either remove `method` or reference it from the JS code.

Comment (etiennestuder) [reply src/main/resources/buildServerResources/admin/buildScanAdminPage.jsp:12]: `onsubmit` is empty. Should it say e.preventDefaults() instead?

Comment (etiennestuder) [inline src/main/resources/buildServerResources/admin/buildScanAdminPage.jsp:12]: Do we need the `<form>` declaration at all?

Comment (etiennestuder) [inline src/main/resources/buildServerResources/admin/buildScanAdminPage.jsp:12]: `onsubmit` is empty. Should it say e.preventDefault() instead?

File: src/main/resources/buildServerResources/admin/buildScanAdminPage.jsp [hunk 9]
```diff
@@ -0,0 +1,23 @@
...
+                <input class="btn btn_mini" type="button" id="startBuildScanCleanupButton" value="Start clean-up now" onclick="return BS.BuildScan.startCleanup();">
+                </form></td><td><forms:progressRing id="progressRing" className="progressRingInline" style="visibility:hidden;"/></td></tr>
+                </table>
```
Comment (etiennestuder) [inline src/main/resources/buildServerResources/admin/buildScanAdminPage.jsp:15]: can you put each `<td>` on its own row? I find it quite hard to right right now.

File: src/main/resources/buildServerResources/js/buildScan.js [hunk 10]
```diff
@@ -0,0 +1,23 @@
...
+        $('startBuildScanCleanupButton').disabled = true
+        $('cleanupResult').textContent = 'Cleanup is running...';
+        $('progressRing').style.visibility = 'visible'
```
Comment (etiennestuder) [inline src/main/resources/buildServerResources/js/buildScan.js:8]: Flip the two lines above (just cosmetics)

**Comment by etiennestuder** [comment]
Thanks for the PR, @jonnybbb . I've merged it into a [branch](https://github.com/etiennestuder/teamcity-build-scan-plugin/tree/jonnybbb-cleanup_custom_storage) and did some polishing.

If you can address the straight-forward items before the meeting, that would be great. We can discuss the remaining questions in the meeting.

File: src/main/java/nu/studer/teamcity/buildscan/internal/cleanup/CustomDataStorageBuildScanCleaner.java [hunk 11]
```diff
@@ -0,0 +1,88 @@
...
+    public CustomDataStorageBuildScanCleaner(SBuildServer buildServer) {
+        server = buildServer;
+    }
+
+    public Integer count() {
+        return server.getSQLRunner().runSql(connection -> {
+            try (Statement statement = connection.createStatement()) {
+                ResultSet resultSet = statement.executeQuery("select count(data_id) from custom_data where data_key='" + BUILD_SCAN_STORAGE_ID + "'");
+                resultSet.next();
+                return resultSet.getInt("count");
```
Comment (etiennestuder) [inline src/main/java/nu/studer/teamcity/buildscan/internal/cleanup/CustomDataStorageBuildScanCleaner.java:29]: This did not work for me locally unless I defined an alias for the count (which I did, as you can see on the merged branch).

File: src/main/java/nu/studer/teamcity/buildscan/internal/cleanup/CustomDataStorageBuildScanCleaner.java (hunk 2 above)
Comment (jonnybbb) [reply src/main/java/nu/studer/teamcity/buildscan/internal/cleanup/CustomDataStorageBuildScanCleaner.java:15]: The majority of rows to be deleted are in the subtable `custom_data_body` - this is just grouping 10 build ids together 1 batch. The actual effected rows in the subtable is very varying, that's why I came up with a defensive low number.

Comment (etiennestuder) [reply src/main/java/nu/studer/teamcity/buildscan/internal/cleanup/CustomDataStorageBuildScanCleaner.java:15]: 👍

File: src/main/resources/buildServerResources/admin/buildScanAdminPage.jsp (hunk 9 above)
Comment (jonnybbb) [reply src/main/resources/buildServerResources/admin/buildScanAdminPage.jsp:21]: done

File: src/main/resources/buildServerResources/admin/buildScanAdminPage.jsp (hunk 8 above)
Comment (jonnybbb) [reply src/main/resources/buildServerResources/admin/buildScanAdminPage.jsp]: return false should achieve the same, right?

Comment (jonnybbb) [reply src/main/resources/buildServerResources/admin/buildScanAdminPage.jsp]: removed

Comment (jonnybbb) [reply src/main/resources/buildServerResources/admin/buildScanAdminPage.jsp]: removed form

File: src/main/resources/buildServerResources/js/buildScan.js (hunk 7 above)
Comment (jonnybbb) [reply src/main/resources/buildServerResources/js/buildScan.js]: done

File: src/main/java/nu/studer/teamcity/buildscan/internal/cleanup/BuildScanAdminPage.java (hunk 3 above)
Comment (jonnybbb) [reply src/main/java/nu/studer/teamcity/buildscan/internal/cleanup/BuildScanAdminPage.java]: done

Comment (jonnybbb) [reply src/main/java/nu/studer/teamcity/buildscan/internal/cleanup/BuildScanAdminPage.java]: yes. We should probably check if it works as well if TC is not deployed under the ROOT context.

File: src/main/resources/buildServerResources/admin/buildScanAdminPage.jsp (hunk 5 above)
Comment (jonnybbb) [reply src/main/resources/buildServerResources/admin/buildScanAdminPage.jsp]: Wouldn't it bet better to rely in the cleanup setting page?

File: src/main/resources/buildServerResources/js/buildScan.js (hunk 10 above)
Comment (jonnybbb) [reply src/main/resources/buildServerResources/js/buildScan.js]: done

**Comment by etiennestuder** [comment]
Cherry-picked the latest changes and fixed all the merge conflicts I got.

**Pull request closed without merging**